src/Connection.ts

```diff
@@ -1,1488 +1,1884 @@
+
+			// Copy the _rev property (whatever that is)
+			if ((obj as any)._rev && this.objects[id]) {
+				(this.objects[id] as any)._rev = (obj as any)._rev;
+			}
+
+			// Detect if there was a change
+			if (
+				!this.objects[id] ||
+				JSON.stringify(this.objects[id]) !== JSON.stringify(obj)
+			) {
+				this.objects[id] = obj;
+				changed = true;
+			}
+		} else if (this.objects[id]) {
+			// The object was deleted
+			delete this.objects[id];
+			changed = true;
+		}
+
+		// Notify all subscribed listeners
+		for (const [_id, sub] of Object.entries(this.objectsSubscribes)) {
+			if (_id === id || sub.reg.test(id)) {
+				sub.cbs.forEach((cb) => cb(id, obj, oldObj));
+			}
+		}
+
+		// Notify the default listener on change
+		if (changed) this.props.onObjectChange?.(id, obj);
+	}
+
+	/**
+	 * Called internally.
+	 * @param id
+	 * @param state
+	 */
+	private stateChange(id: string, state: ioBroker.State | null | undefined) {
+		for (const sub of Object.values(this.statesSubscribes)) {
+			if (sub.reg.test(id)) {
+				for (const cb of sub.cbs) {
+					// TODO: This might not be correct - check what happens if state is a binary state
+					cb(id, (state ?? null) as any);
+				}
+			}
+		}
+	}
+
+	/** Requests data from the server or reads it from the cache */
+	protected async request<T>({
+		cacheKey,
+		forceUpdate,
+		commandTimeout,
+		onTimeout,
+		requireAdmin,
+		requireFeatures,
+		// requestName,
+		executor,
+	}: RequestOptions<T>): Promise<T> {
+		// TODO: mention requestName in errors
+
+		// If the command requires the admin adapter, enforce it
+		if (requireAdmin && Connection.isWeb()) {
+			return Promise.reject(ERRORS.NOT_ADMIN);
+		}
+
+		// Return the cached value if allowed
+		if (cacheKey && !forceUpdate && cacheKey in this._promises) {
+			return this._promises[cacheKey];
+		}
+
+		// Require the socket to be connected
+		if (!this.connected) {
+			return Promise.reject(ERRORS.NOT_CONNECTED);
+		}
+
+		// Check if all required features are supported
+		if (requireFeatures?.length) {
+			for (const feature of requireFeatures) {
+				if (!(await this.checkFeatureSupported(feature))) {
+					throw ERRORS.NOT_SUPPORTED;
+				}
+			}
+		}
+
+		const promise = new Promise<T>(async (resolve, reject) => {
+			const timeoutControl = {
+				elapsed: false,
+				clearTimeout: () => {
+					// no-op unless there is a timeout
+				},
+			};
+			let timeout: NodeJS.Timeout | undefined;
+			if (commandTimeout !== false) {
+				timeout = setTimeout(() => {
+					timeoutControl.elapsed = true;
+					// Let the caller know that the timeout elapsed
+					onTimeout?.();
+					reject(ERRORS.TIMEOUT);
+				}, commandTimeout ?? this.props.cmdTimeout);
+				timeoutControl.clearTimeout = () => {
+					clearTimeout(timeout!);
+				};
+			}
+			// Call the actual function - awaiting it allows us to catch sync and async errors
+			// no matter if the executor is async or not
+			try {
+				await executor(resolve, reject, timeoutControl);
```

Comment:
In tons of cases we need to clear the timeout at the end of the executor. In most cases, it is the last statement. It would be better to add the `clearTimeout()` call to a `finally` clause in this method and only use `clearTimeout()` in special cases.

Reply:
I wanted to avoid a situation where the timeout elapses while a long running executor is still being executed - but I guess these are the special cases you mean?

Reply:
> I wanted to avoid a situation where the timeout elapses while a long running executor is still being executed 

Isn't that the idea of a timeout?

And yes, in certain cases we could clear the timeout before, if we don't need it anymore.